pkg/systemd/quadlet/unitdirs.go

```diff
@@ -0,0 +1,247 @@
+package quadlet
+
+import (
+	"errors"
+	"io/fs"
+	"os"
+	"os/user"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/containers/podman/v5/pkg/logiface"
+)
+
+// This returns whether a file has an extension recognized as a valid Quadlet unit type.
+func IsExtSupported(filename string) bool {
+	ext := filepath.Ext(filename)
+	_, ok := SupportedExtensions[ext]
+	return ok
+}
+
+// This returns default install paths for .Quadlet files for `rootless` and `root` user.
+// Defaults to `/etc/containers/systemd` for root user and `$XDG_CONFIG_HOME/containers/systemd`
+// for rootless users.
+func GetInstallUnitDirPath(rootless bool) string {
+	if rootless {
+		configDir, err := os.UserConfigDir()
+		if err != nil {
+			logiface.Errorf("Warning: %v", err)
+			return ""
+		}
+		return path.Join(configDir, "containers/systemd")
+	}
+	return UnitDirAdmin
+}
+
+// This returns the directories where we read quadlet .container and .volumes from
+// For system generators these are in /usr/share/containers/systemd (for distro files)
+// and /etc/containers/systemd (for sysadmin files).
+// For user generators these can live in $XDG_RUNTIME_DIR/containers/systemd, /etc/containers/systemd/users, /etc/containers/systemd/users/$UID, and $XDG_CONFIG_HOME/containers/systemd
+func GetUnitDirs(rootless bool) []string {
+	paths := NewSearchPaths()
+
+	// Allow overriding source dir, this is mainly for the CI tests
+	if getDirsFromEnv(paths) {
+		return paths.sorted
+	}
+
+	resolvedUnitDirAdminUser := ResolveUnitDirAdminUser()
+	userLevelFilter := GetUserLevelFilter(resolvedUnitDirAdminUser)
+
+	if rootless {
+		systemUserDirLevel := len(strings.Split(resolvedUnitDirAdminUser, string(os.PathSeparator)))
+		nonNumericFilter := GetNonNumericFilter(resolvedUnitDirAdminUser, systemUserDirLevel)
+		getRootlessDirs(paths, nonNumericFilter, userLevelFilter)
+	} else {
+		getRootDirs(paths, userLevelFilter)
+	}
+	return paths.sorted
+}
+
+type searchPaths struct {
+	sorted []string
+	// map to store paths so we can quickly check if we saw them already and not loop in case of symlinks
+	visitedDirs map[string]struct{}
+}
+
+func NewSearchPaths() *searchPaths {
+	return &searchPaths{
+		sorted:      make([]string, 0),
+		visitedDirs: make(map[string]struct{}, 0),
+	}
+}
+
+func (s *searchPaths) Add(path string) {
+	s.sorted = append(s.sorted, path)
+	s.visitedDirs[path] = struct{}{}
+}
+
+func (s *searchPaths) GetSortedPaths() []string {
+	return s.sorted
+}
+
+func (s *searchPaths) Visited(path string) bool {
+	_, visited := s.visitedDirs[path]
+	return visited
+}
+
+func getDirsFromEnv(paths *searchPaths) bool {
+	unitDirsEnv := os.Getenv("QUADLET_UNIT_DIRS")
+	if len(unitDirsEnv) == 0 {
+		return false
+	}
+
+	for _, eachUnitDir := range strings.Split(unitDirsEnv, ":") {
+		if !filepath.IsAbs(eachUnitDir) {
+			logiface.Errorf("%s not a valid file path", eachUnitDir)
```

Comment:
logiface is never initialized for quadlet which means all the log lines are now silenced for quadlet which is wrong.